### PR TITLE
Iliaeg/manualEigen gmm & ba

### PR DESCRIPTION
**Pull request by iliaeg**
-Implemented gmm and ba part of manual eigen module. Added tests for them.
Also proposed a new way of naming tests.

### CODE REVIEW

File: src/cpp/modules/manualEigen/gmm_d.cpp [hunk 1]
```diff
@@ -0,0 +1,437 @@
...
+  {
+    logsumexp_partial_d[max_elem] -= semx;
+    for (int i = 0; i < n; i++)
+      logsumexp_partial_d[i] /= semx;
+  }
+  logsumexp_partial_d[max_elem] += 1.;
+  return log(semx) + mx;
+}
+
+//#ifndef DO_EIGEN
```
Comment (awf) [inline src/cpp/modules/manualEigen/gmm_d.cpp]: Is this commmented-out code now redundant?
Even if not completely, consider a single comment with a point to github repo and hash?

File: src/cpp/modules/manualEigen/gmm_d.cpp [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+      wishart.gamma*wishart.gamma*
+      Map<const VectorXd>(&icf[ik*icf_sz + p], icf_sz - p);
+  }
+
+  return log_wishart_prior(p, k, wishart, sum_qs, Qs, icf);
+}
+
+//#endif
+
+//#if defined DO_EIGEN
```
Comment (awf) [inline src/cpp/modules/manualEigen/gmm_d.cpp]: Again now might be the time to lose these

File: src/cpp/modules/manualEigen/gmm_d.cpp [hunk 3]
```diff
@@ -0,0 +1,437 @@
...
+  vector<MatrixXd> Qs;
+  preprocess(d, k, means, icf, mus, sum_qs, Qs);
+
+  Map<const ArrayXd> map_alphas(alphas, k);
+  gmm_objective_no_priors_d(d, k, n, map_alphas, mus, sum_qs,
+    Qs, x, wishart, err, J);
+  *err += log_wishart_prior_d(d, k, wishart, sum_qs, Qs, icf, J);
+}
+
+//#elif DO_EIGEN_VECTOR
```
Comment (awf) [inline src/cpp/modules/manualEigen/gmm_d.cpp]: OTOH this is another test config we shoud revive.  Not urgent but TTD.

File: src/cpp/shared/gmm_eigen.h [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+  vector<MatrixX<T>> Qs;
+  preprocess(d, k, means, icf, mus, sum_qs, Qs);
+
+  Map<const ArrayX<T>> map_alphas(alphas, k);
+  gmm_objective_no_priors(d, k, n, map_alphas, mus, sum_qs,
+    Qs, x, wishart, err);
+  *err = *err + log_wishart_prior(d, k, wishart, sum_qs, Qs, icf);
+}
+
+#if defined DO_EIGEN_VECTOR //#elif defined DO_EIGEN_VECTOR
```
Comment (awf) [inline src/cpp/shared/gmm_eigen.h]: And this not commented?  Perhaps best to keep consistent (i.e. comment) before it's moved?

File: test/cpp/modules/CMakeLists.txt [hunk 5]
```diff
@@ -1,3 +1,4 @@
 project("ModuleTests")
 
 add_subdirectory ("manual")
+add_subdirectory ("manualEigen")
```
Comment (awf) [inline test/cpp/modules/CMakeLists.txt]: And ultimately add "ManualEigenVectorized

File: test/cpp/modules/manual/GmmTests.cpp [hunk 6]
```diff
@@ -2,7 +2,7 @@
 #include "../../../../src/cpp/runner/ModuleLoader.h"
```
Comment (awf) [inline test/cpp/modules/manual/GmmTests.cpp]: omit ../../src/cpp in middle

File: src/cpp/modules/manualEigen/ba_d.cpp [hunk 7]
```diff
@@ -0,0 +1,387 @@
...
+    proj[1] = proj[1] * f + x0[1];
+}
+
+// Arguments:
+// cam - double[BA_NCAMPARAMS]
+// X - double[3]
+// w, feat_x, feat_y - double
+// err - double[2]
+// J - 2 x (BA_NCAMPARAMS+3+1) in column major
+void compute_reproj_error_d(
```
Comment (msdmkats) [inline src/cpp/modules/manualEigen/ba_d.cpp]: This function is a copy of one from manual/ba_d.cpp and it does not use Eigen in the computation. In fact, the code in this file that uses Eigen doesn't seem to be called from anywhere.

File: test/cpp/modules/manual/GmmTests.cpp (hunk 6 above)
Comment (msdmkats) [reply test/cpp/modules/manual/GmmTests.cpp]: Can't do that - this file lies in the /test/cpp/... folder, not in /src/cpp/...

File: src/cpp/modules/manualEigen/gmm_d.cpp (hunk 1 above)
Comment (msdmkats) [reply src/cpp/modules/manualEigen/gmm_d.cpp]: It is. This code is now in /src/cpp/modules/manual/gmm.cpp

File: tools/cpp-common/gmm_eigen.h [hunk 8]
```diff
@@ -1,9 +1,12 @@
 #include <cmath>
```
Comment (msdmkats) [inline tools/cpp-common/gmm_eigen.h]: Isn't this file supposed to be moved to /src/cpp/shared? We should delete it from here, then. Need to check that all the references in the codebase are corrected before doing that, though.

Comment (iliaeg) [reply tools/cpp-common/gmm_eigen.h]: I prefer not to delete gmm_eigen.h until eigen vectorized will be implemented.

File: test/cpp/modules/manualEigen/BaTests.cpp [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    module->calculateObjective(1);
+    module->calculateJacobian(1);
+    auto output = module->output();
+    EXPECT_EQ(30, output.J.nrows);
+    EXPECT_EQ(62, output.J.ncols);
+    EXPECT_EQ(31, output.J.rows.size());
+    EXPECT_EQ(310, output.J.cols.size());
+    EXPECT_EQ(310, output.J.vals.size());
+    EXPECT_NEAR(228.877, output.J.vals[0], 0.001);
+    EXPECT_NEAR(634.575, output.J.vals[1], 0.001);
```
[the remark this answers is not in the document]
Comment (iliaeg) [reply test/cpp/modules/manualEigen/BaTests.cpp]: Yes, it's a great feature, Dmitry has implemented this part for manual ba few days ago.

**Pull request merged**